README
= Description
   The DBI package is a vendor independent interface for accessing databases.
   It is similar, but not identical to, Perl's DBI module.
   
= Prerequisites  
   Ruby 1.8.0 or later.

= RubyForge Project
   General information: http://rubyforge.org/projects/ruby-dbi/
   Downloads: http://rubyforge.org/frs/?group_id=234

= Installation
   Many available DBDs come with this package, but you should only install
   the DBDs you really need.

   To install everything:

   ruby setup.rb config
   ruby setup.rb setup
   ruby setup.rb install

   To install dbi and some DBDs:

   ruby setup.rb config --with=dbi,dbd_pg, ...
   ruby setup.rb setup
   ruby setup.rb install

   Choose the packages to install by specifing them after the option --with.

= Synopsis
   require 'dbi'

   # Connect to a database, old style
   dbh = DBI.connect('DBI:Mysql:test', 'testuser', 'testpwd')

   # Insert some rows, use placeholders
   1.upto(13) do |i|
      sql = "insert into simple01 (SongName, SongLength_s) VALUES (?, ?)"
      dbh.do(sql, "Song #{i}", "#{i*10}")
   end 

   # Select all rows from simple01
   sth = dbh.prepare('select * from simple01')
   sth.execute

   # Print out each row
   while row=sth.fetch do
      p row
   end

   # Close the statement handle when done
   sth.finish

   # Don't prepare, just do it!
   dbh.do('delete from simple01 where internal_id > 10')

   # And finally, disconnect
   dbh.disconnect

   # Same example, but a little more Ruby-ish
   DBI.connect('DBI:Mysql:test', 'testuser', 'testpwd') do | dbh |

      sql = "insert into simple01 (SongName, SongLength_s) VALUES (?, ?)"

      dbh.prepare(sql) do | sth | 
         1.upto(13) { |i| sth.execute("Song #{i}", "#{i*10}") }
      end 

      dbh.select_all('select * from simple01') do | row |
         p row
      end

      dbh.do('delete from simple01 where internal_id > 10')

   end

= Available Database Drivers (DBDs)
== DBD::ADO
   ADO (ActiveX Data Objects)
   Requires win32ole, part of the Ruby standard library.

== DBD::DB2
   DB2
   Depends on the ruby-db2 package, available on the RAA.

== DBD::Frontbase
   Frontbase
   Depends on the ruby-frontbase package, available on the RAA.

== DBD::InterBase
   Interbase
   Depends on the interbase package, available on the RAA.

== DBD::mSQL
   mSQL
   Depends on the msql package, available on the RAA.

== DBD::MySQL
   MySQL
   Depends on the mysql-ruby package from http://www.tmtm.org/mysql or
   available from the RAA.

== DBD::ODBC
   ODBC
   Depends on the ruby-odbc package (0.5 or later, 0.9.3 or later recommended) at
   http://www.ch-werner.de/rubyodbc or available from the RAA.  Works together
   with unix-odbc.

== DBD::Oracle
   Oracle
   Depends on the oracle package 0.2.11 or later, available on the RAA.  This
   driver is deprecated in favor of DBD::OCI8.

== DBD::OCI8
   OCI8 (Oracle)
   Depends on the the ruby-oci8 package, available on the RAA and RubyForge.

== DBD::Pg
   PostgreSQL
   Depends on either the postgres or postgres-pr package, both available on
   the RAA.

== DBD::Proxy
   Proxy/Server
   Depends on distributed Ruby (DRb), part of the Ruby standard library.

== DBD::SQLite
   SQLite (versions 2.x and earlier)
   Depends on the sqlite-ruby package, available on rubyforge.

== DBD::SQLite
   SQLite 3.x
   Depends on the sqlite3-ruby package, available on rubyforge.

== DBD::SQLRelay
   SQLRelay
   Depends on the sqlrelay package at http://www.firstworks.com/sqlrelay

= Additional Documentation
   See the directories lib/*/doc or ext/*/doc for DBI and DBD specific information.
   The DBI specification is at lib/dbi/doc/DBI_SPEC.
   The DBD specification is at lib/dbi/doc/DBD_SPEC.

= Articles
== Using the Ruby DBI Module
   http://www.kitebird.com/articles/ruby-dbi.html
   
= Applications
== sqlsh.rb
   The SQL command line interpreter sqlsh.rb is available in directory
   bin/commandline. It gets installed by default.

= License

   Copyright (c) 2008 Erik Hollensbe and Christopher Maujean.

   Copyright (c) 2005-2006 Kirk Haines, Francis Hwang, Patrick May and Daniel
   Berger.

   Copyright (c) 2001, 2002, 2003, 2004 Michael Neumann <mneumann@ntecs.de>
   and others (see the beginning of each file for copyright holder information).

   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.
   3. The name of the author may not be used to endorse or promote products
      derived from this software without specific prior written permission.

   THIS SOFTWARE IS PROVIDED 'AS IS' AND ANY EXPRESS OR IMPLIED WARRANTIES,
   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
   AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
   THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
   OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
   OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
   ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

   This is the BSD license which is less restrictive than GNU's GPL
   (General Public License).

= Contributors

   Jun Mukai <mukai@jmuk.org>
      Contributed initial SQLite3 DBD.

   John J. Fox IV
      Lots of help testing on multiple platforms.

   Kirk Haines
      One of the authors of the rewrite effort (January 2006).

   Francis Hwang
      One of the authors of the rewrite effort (January 2006).

   Patrick May
      One of the authors of the rewrite effort (January 2006).

   Daniel Berger
      One of the authors of the rewrite effort (January 2006).

   Michael Neumann
      Original author of Ruby/DBI; wrote the DBI and most of the DBDs.

   Rainer Perl
      Author of Ruby/DBI 0.0.4 from which many good ideas were taken.

   Jim Weirich
      Original author of DBD::Pg.  Wrote additional code (e.g. sql.rb,
      testcases).  Gave many helpful hints and comments. 

   Eli Green
      Implemented DatabaseHandle#columns for Mysql and Pg. 

   Masatoshi SEKI
      For his version of module BasicQuote in sql.rb. 

   John Gorman
      For his case insensitive load_driver patch and parameter parser. 

   David Muse
      For testing the DBD::SQLRelay and for his initial DBD. 

   Jim Menard
      Extended DBD::Oracle for method columns. 

   Joseph McDonald
      Fixed bug in DBD::Pg (default values in method columns). 

   Norbert Gawor
      Fixed bug in DBD::ODBC (method columns) and proxyserver. 

   James F. Hranicky
      Patch for DBD::Pg (cache PGResult#result in Tuples) which increased
      performance by a factor around 100. 

   Stephen Davies
      Added method Statement#fetch_scroll for DBD::Pg. 

   Dave Thomas
      Several enhancements. 

   Brad Hilton
      Column coercing patch for DBD::Mysql. 

   Sean Chittenden
      Originally a co-owner of the project.  Submitted several patches
      and helped with lots of comments.

   MoonWolf
      Provided the quote/escape_byte patch for DBD::Pg, DBD::SQLite patch and
      Database#columns implementation. Further patches. 

   Paul DuBois
      Fixed typos and formatting. Maintains DBD::Mysql. 

   Tim Bates
      Bug fixes for Mysql and DBI.

   Brian Candler
      Zero-padding date/time/timestamps fix. 

   Florian G. Pflug

ChangeLog
== 0.2.0 - 03/16/2008

=== Summary

* SQLite (2.x) driver rewritten in ruby
* SQLite (3.x) driver provided by Jun Mukai
* Migrated DBD::Pg to 'pg' low-level library
* bytea and LOB support for DBD::Pg
* New DBD Test suite
* Tons of bug fixes

commit:: 7bc29dd74badd75daa99b4c1f2c97e93dbc6a405
Author:: Erik Hollensbe <erik@hollensbe.org>
Date::   Fri Mar 14 11:13:48 2008 -0700

* Fixed two bugs in the Pg driver (thanks again Mike Pomraning):
* A tables bug where the system tables were not being sent back.
* A columns bug where the columns were not properly selected when the schema search path was altered.

commit:: eaf228bdf1c992cd8c523872057e5d2e1901eb6d
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Wed Mar 12 15:35:08 2008 -0700

* Fix for timestamp type handling under mysql

commit:: ce4c59e0a60f0dc4bbd66cc84c1b2a50df17c6ca
Author:: Erik Hollensbe <erik@hollensbe.org>
Date::   Tue Mar 11 10:58:18 2008 -0700

* Some fixes for timestamp handling for SQLite3

commit:: beb4c5b31b355180761df024c6933e1b67a2f59c
Author:: Erik Hollensbe <erik@hollensbe.org>
Date::   Tue Mar 11 09:09:37 2008 -0700

* Patch provided by Mike Pomraning that gets timestamps in a sane fashion!

commit:: a6d70ffb23bb0493cd73ebe38dd413a9d86173bd
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sat Mar 8 19:05:58 2008 -0800

* Fixed a DBI::Timestamp#to_s error

commit:: 48690de14c144ac030b12a2a627db611a0ca257c
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sat Mar 8 01:19:42 2008 -0800

* Fix for #3916 - sth.any? and sth.rows.zero? not consistent across DBDs

commit:: 55f118d3e54dc546b133083271d80caccf00ab53
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sat Mar 8 00:34:51 2008 -0800

* Peter Schiller's thread-safety patch
* Some more tests surrounding the row count and fetchable?

commit:: b496e0d981891fad6291dc49161963a15f05472a
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Thu Mar 6 22:55:07 2008 -0800

* Plenty of DBI::Time fixes (based on a patch from #7710 and tests that were written as generals)

commit:: 679bde33924ee368322f7ece4509822e07e68955
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Thu Mar 6 21:42:00 2008 -0800

* #6965 Fix and test (not ideal, but it does work)

commit:: 3819bd468b941465abaffd6b7adc5afa1ddd0406
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Thu Mar 6 21:11:06 2008 -0800

* #4727 integrated

commit:: 32c0e5e6fea308cd52891bf6dac589ad84b29ba4
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Thu Mar 6 20:57:00 2008 -0800

* Applied patch from ticket #7711

commit:: 34286112df17fbdb7588f45b176a42fb3a77aaae
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sun Feb 24 12:56:35 2008 -0800

* Fixed bug #10466

commit:: 1f19663fd9b31b0f2f0a111f227220d5ee8cbe1d
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sun Feb 24 12:09:44 2008 -0800

* Fixed #1082

commit:: f3e6e077e4073c892d8449ec6e1912ffe9f76f95
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sun Feb 24 12:02:23 2008 -0800

* ByteA fixes!

commit:: 8668eb0c1561b49b144f4a56c3edc2bce90bc2a0
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sun Feb 24 00:29:00 2008 -0800

* Fixed most of the LOB support (yay!)

commit:: 32ac057631b16429051c90fa9aab3ed1bb560e32
Author:: Erik Hollensbe <erikh@speyside.(none)>
Date::   Sat Feb 23 23:35:19 2008 -0800

* Initial porting attempt from ruby-postgres to ruby-pg
* Test fixes and additions

== 0.1.2 - 02-Sep-2006

=== General
* Paul DuBois: test/dbi/tc_dbi.rb: DBI version number test failed because it
  was expected to be 0.1.0. (Should have been 0.1.1.) But 0.1.1 has been
  released, so update DBI version number and expected number in test to 0.1.2.

== 0.1.1 - 12-Aug-2006

=== General

* Paul DuBois: README: Added URLs for the RubyForge DBI project.
* Patrick May: removed misleading space in setup.rb config example.
* Francis Hwang: fixed bug 597: behaviour of StatementHandle#fetch doesn't
  match DBI specification

lib/dbd/Mysql.rb
#
# DBD::Mysql
#
# Copyright (c) 2001, 2002 Michael Neumann <neumann@s-direktnet.de>
# 
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions 
# are met:
# 1. Redistributions of source code must retain the above copyright 
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright 
#    notice, this list of conditions and the following disclaimer in the 
#    documentation and/or other materials provided with the distribution.
# 3. The name of the author may not be used to endorse or promote products
#    derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
# AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
# THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# $Id$
#

require "mysql"
require "thread"   # for Mutex


# The MySQL-specific column_info method defines some MySQL-specific
# members of the column info attribute array. Extend ColumnInfo by
# adding methods so that info["mysql_x"], for each MySQL-specific x, also
# has get and set info.mysql_x and info.mysql_x= methods.

class ColumnInfo

  # Get the column's MySQL type code
  def mysql_type
     self['mysql_type']
  end
   
  # Set the column's MySQL type code
  def mysql_type=(val)
     self['mysql_type'] = val
  end

  # Get the column's MySQL type name
  def mysql_type_name
     self['mysql_type_name']
  end
   
  # Set the column's MySQL type name
  def mysql_type_name=(val)
     self['mysql_type_name'] = val
  end

  # Get the column's MySQL length
  def mysql_length
     self['mysql_length']
  end
   
  # Set the column's MySQL length
  def mysql_length=(val)
     self['mysql_length'] = val
  end

  # Get the column's MySQL max length
  def mysql_max_length
     self['mysql_max_length']
  end
   
  # Set the column's MySQL max length
  def mysql_max_length=(val)
     self['mysql_max_length'] = val
  end

  # Get the column's MySQL flags
  def mysql_flags
     self['mysql_flags']
  end
   
  # Set the column's MySQL flags
  def mysql_flags=(val)
     self['mysql_flags'] = val
  end

end # class ColumnInfo

module DBI
module DBD
module Mysql

VERSION          = "0.3.3"
USED_DBD_VERSION = "0.2"

MyError = ::MysqlError

  module Util

    private

    # Raise exception using information from MysqlError object e.
    # For state value, use SQLSTATE value if mysql-ruby defines
    # sqlstate method, otherwise nil.

    def error(e)
      sqlstate = e.respond_to?("sqlstate") ? e.sqlstate : nil
      raise DBI::DatabaseError.new(e.message, e.errno, sqlstate)
    end

  end # module Util

class Driver < DBI::BaseDriver
  include Util

  def initialize
    super(USED_DBD_VERSION)
  end

  def default_user
    ['', nil]
  end

  def connect(dbname, user, auth, attr)
    # connect to database server
    hash = Utils.parse_params(dbname)

    hash['host'] ||= 'localhost'

    # these two connection parameters should be passed as numbers
    hash['port'] = hash['port'].to_i unless hash['port'].nil?
    hash['flag'] = hash['flag'].nil? ? 0 : hash['flag'] = hash['flag'].to_i

    handle = ::Mysql.init

    # Look for options in connect string to be handled
    # through mysql_options() before connecting
    !hash['mysql_read_default_file'].nil? and
      handle.options(::Mysql::READ_DEFAULT_FILE,
                     hash['mysql_read_default_file'])
    !hash['mysql_read_default_group'].nil? and
      handle.options(::Mysql::READ_DEFAULT_GROUP,
                     hash['mysql_read_default_group'])
    # The following options can be handled either using mysql_options()
    # or in the flag argument to connect().
    hash['mysql_compression'].to_i != 0 and
      handle.options(::Mysql::OPT_COMPRESS, nil)
    hash['mysql_local_infile'].to_i != 0 and
      handle.options(::Mysql::OPT_LOCAL_INFILE, true)

    # Look for options to be handled in the flags argument to connect()
    if !hash['mysql_client_found_rows'].nil?
      if hash['mysql_client_found_rows'].to_i != 0
        hash['flag'] |= ::Mysql::CLIENT_FOUND_ROWS
      else
        hash['flag'] &= ~::Mysql::CLIENT_FOUND_ROWS
      end
    end

    handle.connect(hash['host'], user, auth, hash['database'], hash['port'], hash['socket'], hash['flag'])

    return Database.new(handle, attr)
  rescue MyError => err
    error(err)
  end

  def data_sources
    handle = ::Mysql.new
    res = handle.list_dbs.collect {|db| "dbi:Mysql:database=#{db}" }
    handle.close
    return res
  rescue MyError => err
    error(err)
  end

  # Driver-specific functions ------------------------------------------------

  public

  def __createdb(db, host, user, password, port=nil, sock=nil, flag=nil)
    handle = ::Mysql.connect(host, user, password, nil, port, sock, flag)
    begin
      handle.create_db(db)
    ensure
      handle.close if handle
    end
  end

  def __dropdb(db, host, user, password, port=nil, sock=nil, flag=nil)
    handle = ::Mysql.connect(host, user, password, nil, port, sock, flag)
    begin
      handle.drop_db(db)
    ensure
      handle.close if handle
    end
  end

  def __shutdown(host, user, password, port=nil, sock=nil, flag=nil)
    handle = ::Mysql.connect(host, user, password, nil, port, sock, flag)
    begin
      handle.shutdown
    ensure
      handle.close if handle
    end
  end

  def __reload(host, user, password, port=nil, sock=nil, flag=nil)
    handle = ::Mysql.connect(host, user, password, nil, port, sock, flag)
    begin
      handle.reload
    ensure
      handle.close if handle
    end
  end

end # class Driver

class Database < DBI::BaseDatabase
  include Util
  include SQL::BasicBind

  # Eli Green:
  #   The hope is that we don't ever need to just assume the default values.
  #   However, in some cases (notably floats and doubles), I have seen
  #   "show fields from table" return absolutely zero information about size
  #   and precision. Sigh. I probably should have made a struct to store
  #   this info in ... but I didn't.
  MYSQL_to_XOPEN = {
    "TINYINT"    => [SQL_TINYINT, 1, nil],
    "SMALLINT"   => [SQL_SMALLINT, 6, nil],
    "MEDIUMINT"  => [SQL_SMALLINT, 6, nil],
    "INT"        => [SQL_INTEGER, 11, nil],
    "INTEGER"    => [SQL_INTEGER, 11, nil],
    "BIGINT"     => [SQL_BIGINT, 25, nil],
    "INT24"      => [SQL_BIGINT, 25, nil],
    "REAL"       => [SQL_REAL, 12, nil],
    "FLOAT"      => [SQL_FLOAT, 12, nil],
    "DECIMAL"    => [SQL_DECIMAL, 12, nil],
    "NUMERIC"    => [SQL_NUMERIC, 12, nil],
    "DOUBLE"     => [SQL_DOUBLE, 22, nil],
    "CHAR"       => [SQL_CHAR, 1, nil],
    "VARCHAR"    => [SQL_VARCHAR, 255, nil],
    "DATE"       => [SQL_DATE, 10, nil],
    "TIME"       => [SQL_TIME, 8, nil],
    "TIMESTAMP"  => [SQL_TIMESTAMP, 19, nil],
    "DATETIME"   => [SQL_TIMESTAMP, 19, nil],
    "TINYBLOB"   => [SQL_BINARY, 255, nil],
    "BLOB"       => [SQL_VARBINARY, 65535, nil],
    "MEDIUMBLOB" => [SQL_VARBINARY, 16277215, nil],
    "LONGBLOB"   => [SQL_LONGVARBINARY, 2147483657, nil],
    "TINYTEXT"   => [SQL_VARCHAR, 255, nil],
    "TEXT"       => [SQL_LONGVARCHAR, 65535, nil],
    "MEDIUMTEXT" => [SQL_LONGVARCHAR, 16277215, nil],
    "LONGTEXT"   => [SQL_LONGVARCHAR, 2147483657, nil],
    "ENUM"       => [SQL_CHAR, 255, nil],
    "SET"        => [SQL_CHAR, 255, nil],
    "BIT"        => [SQL_BIT, 8, nil],
    nil          => [SQL_OTHER, nil, nil]
  }

  # Map MySQL numeric type codes to:
  # - (uppercase) MySQL type names
  # - coercion method

  TYPE_MAP = {}
  ::Mysql::Field.constants.grep(/^TYPE_/).each do |const|
    mysql_type = MysqlField.const_get(const)  # numeric type code
    coercion_method = :as_str                 # default coercion method
    case const
    when 'TYPE_TINY':         mysql_type_name = 'TINYINT'
                              coercion_method = :as_int
    when 'TYPE_SHORT':        mysql_type_name = 'SMALLINT'
                              coercion_method = :as_int
    when 'TYPE_INT24':        mysql_type_name = 'MEDIUMINT'
                              coercion_method = :as_int
    when 'TYPE_LONG':         mysql_type_name = 'INT'
                              coercion_method = :as_int
    when 'TYPE_LONGLONG':     mysql_type_name = 'BIGINT'
                              coercion_method = :as_int
    when 'TYPE_FLOAT':        mysql_type_name = 'FLOAT'
                              coercion_method = :as_float
    when 'TYPE_DOUBLE':       mysql_type_name = 'DOUBLE'
                              coercion_method = :as_float
    when 'TYPE_VAR_STRING',
         'TYPE_STRING':       mysql_type_name = 'VARCHAR'    # questionable?
                              coercion_method = :as_str
    when 'TYPE_DATE':         mysql_type_name = 'DATE'
                              coercion_method = :as_date
    when 'TYPE_TIME':         mysql_type_name = 'TIME'
                              coercion_method = :as_time
    when 'TYPE_DATETIME', 
         'TYPE_TIMESTAMP':    mysql_type_name = 'DATETIME'
                              coercion_method = :as_timestamp
    when 'TYPE_CHAR':         mysql_type_name = 'TINYINT'    # questionable?
    when 'TYPE_TINY_BLOB':    mysql_type_name = 'TINYBLOB'   # questionable?
    when 'TYPE_MEDIUM_BLOB':  mysql_type_name = 'MEDIUMBLOB' # questionable?
    when 'TYPE_LONG_BLOB':    mysql_type_name = 'LONGBLOB'   # questionable?
    when 'TYPE_GEOMETRY':     mysql_type_name = 'BLOB'       # questionable?
    when 'TYPE_YEAR',
         'TYPE_DECIMAL',                                     # questionable?
         'TYPE_BLOB',                                        # questionable?
         'TYPE_ENUM',
         'TYPE_SET',
         'TYPE_BIT',
         'TYPE_NULL':         mysql_type_name = const.sub(/^TYPE_/, '')
    else
                              mysql_type_name = 'UNKNOWN'
    end
    TYPE_MAP[mysql_type] = [mysql_type_name, coercion_method]
  end
  TYPE_MAP[nil] = ['UNKNOWN', :as_str]

  def initialize(handle, attr)
    super
    # check server version to determine transaction capability
    ver_str = @handle.get_server_info
    major, minor, teeny = ver_str.split(".")
    teeny.sub!(/\D*$/, "")  # strip any non-numeric suffix if present
    server_version = major.to_i*10000 + minor.to_i*100 + teeny.to_i
    # It's not until 3.23.17 that SET AUTOCOMMIT,
    # BEGIN, COMMIT, and ROLLBACK all are available
    @have_transactions = (server_version >= 32317)
    # assume that the connection begins in AutoCommit mode
    @attr['AutoCommit'] = true
    @mutex = Mutex.new 
  end

  def disconnect
    self.rollback unless @attr['AutoCommit']
    @handle.close
  rescue MyError => err
    error(err)
  end

  def ping
    begin
      @handle.ping
      return true
    rescue MyError
      return false
    end
  end

  def tables
    @handle.list_tables
  rescue MyError => err
    error(err)
  end

  # Eli Green (fixed up by Michael Neumann)
  def columns(table)
    dbh = DBI::DatabaseHandle.new(self)
    uniques = []
    dbh.execute("SHOW INDEX FROM #{table}") do |sth|
      sth.each do |row|
        uniques << row[4] if row[1] == "0"
      end
    end  

    ret = nil
    dbh.execute("SHOW FIELDS FROM #{table}") do |sth|
      ret = sth.collect do |row|
        name, type, nullable, key, default, extra = row
        #type = row[1]
        #size = type[type.index('(')+1..type.index(')')-1]
        #size = 0
        #type = type[0..type.index('(')-1]

        sqltype, type, size, decimal = mysql_type_info(row[1])
        col = Hash.new
        col['name']           = name
        col['sql_type']       = sqltype
        col['type_name']      = type
        col['nullable']       = nullable == "YES"
        col['indexed']        = key != ""
        col['primary']        = key == "PRI"
        col['unique']         = uniques.index(name) != nil
        col['precision']      = size
        col['scale']          = decimal
        col['default']        = row[4]
        col
      end # collect
    end # execute
   
    ret
  end

  def do(stmt, *bindvars)
    sql = bind(self, stmt, bindvars)
    @mutex.synchronize { 
      @handle.query_with_result = false
      @handle.query(sql)
      @handle.affected_rows     # return value
    }
  rescue MyError => err
    error(err)
  end
 

  def prepare(statement)
    Statement.new(self, @handle, statement, @mutex)
  end

  def commit
    if @have_transactions
      self.do("COMMIT")
    else
      raise NotSupportedError
    end
  rescue MyError => err
    error(err)
  end

  def rollback
    if @have_transactions
      self.do("ROLLBACK")
    else
      raise NotSupportedError
    end
  rescue MyError => err
    error(err)
  end


  def quote(value)
    case value
    when String
      "'#{@handle.quote(value)}'"
    when DBI::Binary
      "'#{@handle.quote(value.to_s)}'"
    when TrueClass
      "'1'"
    when FalseClass
      "'0'"
    else
      super
    end
  end

  def []=(attr, value)
    case attr
    when 'AutoCommit'
      if @have_transactions
        self.do("SET AUTOCOMMIT=" + (value ? "1" : "0"))
      else
        raise NotSupportedError
      end
    else
      raise NotSupportedError
    end
    @attr[attr] = value
  end

  private # -------------------------------------------------

  # Eli Green
  # Parse column type string (from SHOW FIELDS) to extract type info:
  # - sqltype: XOPEN type number
  # - type: MySQL type name
  # - size: column length (or precision)
  # - decimal: number of decimals (scale)
  def mysql_type_info(typedef)
    sqltype, type, size, decimal = nil, nil, nil, nil

    pos = typedef.index('(')
    if not pos.nil?
      type = typedef[0..pos-1]
      size = typedef[pos+1..-2]
      pos = size.index(',')
      if not pos.nil?
        size, decimal = size.split(',', 2)
        decimal = decimal.to_i
      end
      size = size.to_i
    else
      type = typedef
    end

    type_info = MYSQL_to_XOPEN[type.upcase] || MYSQL_to_XOPEN[nil]
    sqltype = type_info[0]
    if size.nil? then size = type_info[1] end
    if decimal.nil? then decimal = type_info[2] end
    return sqltype, type, size, decimal
  end

  
  # Driver-specific functions ------------------------------------------------

  public

  def __createdb(db)
    @handle.create_db(db)
  end

  def __dropdb(db)
    @handle.drop_db(db)
  end

  def __shutdown
    @handle.shutdown
  end

  def __reload
    @handle.reload
  end

  def __insert_id
    @handle.insert_id
  end

  def __thread_id
    @handle.thread_id
  end

  def __info
    @handle.info
  end

  def __host_info
    @handle.host_info
  end

  def __proto_info
    @handle.proto_info
  end

  def __server_info
    @handle.server_info
  end

  def __client_info
    @handle.client_info
  end

  def __client_version
    @handle.client_version
  end

  def __stat
    @handle.stat
  end


end # class Database


class Statement < DBI::BaseStatement
  include Util
  #include SQL::BasicBind

  def initialize(parent, handle, statement, mutex)
    super(nil)

    @parent, @handle, @mutex = parent, handle, mutex
    @params = []

    @prep_stmt = DBI::SQL::PreparedStatement.new(@parent, statement)
  end

  def bind_param(param, value, attribs)
    raise InterfaceError, "only ? parameters supported" unless param.is_a? Fixnum
    @params[param-1] = value 
  end

  def execute
    sql = @prep_stmt.bind(@params)
    @mutex.synchronize {
      @handle.query_with_result = true
      @res_handle = @handle.query(sql)
      @column_info = self.column_info
      @coerce = DBI::SQL::BasicQuote::Coerce.new
      @current_row = 0
      @rows = DBI::SQL.query?(sql) ? 0 : @handle.affected_rows 
    }
  rescue MyError => err
    error(err)
  end

  def finish
    @res_handle.free if @res_handle
  rescue MyError => err
    error(err)
  end

  def fill_array(rowdata)
    return nil if rowdata.nil?
    row = []
    rowdata.each_with_index { |value, index|
      info = @column_info[index]
      type = info['mysql_type']
      type_symbol =
        if SQL_TINYINT == info['sql_type'] and 1 == info['precision']
          :as_bool
        else
          Database::TYPE_MAP[type][1] || :as_str rescue :as_str
        end
      row[index] = @coerce.coerce(type_symbol, value)
    }
    row
  end

  def fetch
    @current_row += 1
    fill_array(@res_handle.fetch_row)
  rescue MyError => err
    error(err)
  end

  def fetch_scroll(direction, offset)
      case direction
      when SQL_FETCH_NEXT
        @current_row += 1
        fill_array(@res_handle.fetch_row)
      when SQL_FETCH_PRIOR
        @res_handle.data_seek(@current_row - 1)
        fill_array(@res_handle.fetch_row)
      when SQL_FETCH_FIRST
        @current_row = 1
        @res_handle.data_seek(@current_row - 1)
        fill_array(@res_handle.fetch_row)
      when SQL_FETCH_LAST
        @current_row = @res_handle.num_rows
        @res_handle.data_seek(@current_row - 1)
        fill_array(@res_handle.fetch_row)
      when SQL_FETCH_ABSOLUTE
        @current_row = offset + 1
        @res_handle.data_seek(@current_row - 1)
        fill_array(@res_handle.fetch_row)
      when SQL_FETCH_RELATIVE
        @current_row += offset + 1
        @res_handle.data_seek(@current_row - 1)
        fill_array(@res_handle.fetch_row)
      else
        raise NotSupportedError
      end
    #end
  end

  def column_info
    retval = []

    return [] if @res_handle.nil?

    unique_key_flag = MysqlField.const_get(:UNIQUE_KEY_FLAG)
    multiple_key_flag = MysqlField.const_get(:MULTIPLE_KEY_FLAG)
    indexed = (unique_key_flag | multiple_key_flag)

    # Note: Cannot get 'default' column attribute because MysqlField.def
    # is set only by mysql_list_fields()

    @res_handle.fetch_fields.each {|col| 
      mysql_type_name = Database::TYPE_MAP[col.type][0] rescue nil
      xopen_info = Database::MYSQL_to_XOPEN[mysql_type_name] ||
                   Database::MYSQL_to_XOPEN[nil]
      sql_type = xopen_info[0]
      type_name = DBI::SQL_TYPE_NAMES[sql_type]

      retval << {
                  # Standard Ruby DBI column attributes
                  'name'        => col.name,
                  'sql_type'    => sql_type,
                  'type_name'   => type_name,
                  'precision'   => col.length,
                  'scale'       => col.decimals,
                  'nullable'    => !col.is_not_null?,
                  'indexed'     => ((col.flags & indexed) != 0) ||
                                   col.is_pri_key?,
                  'primary'     => col.is_pri_key?,
                  'unique'      => ((col.flags & unique_key_flag) != 0) ||
                                   col.is_pri_key?,
                  # MySQL-specific attributes (signified by leading "mysql_")
                  'mysql_type'       => col.type,
                  'mysql_type_name'  => mysql_type_name,
                  'mysql_length'     => col.length,
                  'mysql_max_length' => col.max_length,
                  'mysql_flags'      => col.flags
                }
    }
    retval
  rescue MyError => err
    error(err)
  end

  def rows
    @rows
  end

=begin
  def []=(attr, value)
    case attr
    when 'mysql_use_result'
      @attr['mysql_store_result'] = ! value
      @attr['mysql_use_result']   = value
    when 'mysql_store_result'
      @attr['mysql_use_result']   = ! value
      @attr['mysql_store_result'] = value
    else
      raise NotSupportedError
    end
  end
=end

end # class Statement


end # module Mysql
end # module DBD
end # module DBI
